lib/src/test/java/io/ably/lib/http/HttpTest.java

```diff
@@ -0,0 +1,637 @@
+		http.setHost(hostExpected);
+		ArgumentCaptor<URL> url = ArgumentCaptor.forClass(URL.class);
+
+        /* Partially mock http */
+		doReturn(responseExpected) /* Provide response */
+				.when(http) /* when following method is executed on {@code Http} instance */
+				.httpExecute(
+						url.capture(), /* capture url arguments passed down httpExecute to assert fallback behavior executed with valid rest host */
+						anyString(), /* Ignore */
+						aryEq(new Param[0]), /* Ignore */
+						any(Http.RequestBody.class), /* Ignore */
+						anyBoolean(), /* Ignore */
+						any(Http.ResponseHandler.class) /* Ignore */
+				);
+
+		String responseActual = (String) http.ablyHttpExecute(
+				"", /* Ignore */
+				"", /* Ignore */
+				new Param[0], /* Ignore */
+				new Param[0], /* Ignore */
+				mock(Http.RequestBody.class), /* Ignore */
+				mock(Http.ResponseHandler.class) /* Ignore */
+		);
+
+
+        /* Verify
+		 *   - http call executed once,
+         *   - with given host,
+         *   - and delivered expected response */
+		verify(http, times(1))
+				.httpExecute( /* Just validating call counter. Ignore following parameters */
+						url.capture(), /* Ignore */
+						anyString(), /* Ignore */
+						aryEq(new Param[0]), /* Ignore */
+						any(Http.RequestBody.class), /* Ignore */
+						anyBoolean(), /* Ignore */
+						any(Http.ResponseHandler.class) /* Ignore */
+				);
+
+		assertThat(url.getValue().getHost(), is(equalTo(hostExpected)));
+		assertThat(responseActual, is(equalTo(responseExpected)));
+	}
+
+	/**
+	 * <strong>This method mocks the API behavior</strong>
+	 * <p>
+	 * Validates http is using a fallback host when HostFailedException thrown
+	 * </p>
+	 * <p>
+	 * Spec: RSC15a
+	 * </p>
+	 *
+	 * @throws Exception
+	 */
+	@Test
+	public void http_execute_singlefallback() throws Exception {
+		Http http = Mockito.spy(new Http(new ClientOptions(), null));
+
+		String hostExpectedPattern = PATTERN_HOST_FALLBACK;
+		String responseExpected = "Lorem Ipsum";
+		ArgumentCaptor<URL> url = ArgumentCaptor.forClass(URL.class);
+		http.setHost(Defaults.HOST_REST);
+
+        /* Partially mock http */
+		Answer answer = new GrumpyAnswer(
+				1, /* Throw exception once (1) */
+				AblyException.fromErrorInfo(ErrorInfo.fromResponseStatus("Internal Server Error", 500)), /* That is HostFailedException */
+				responseExpected /* Then return a valid response with the second call */
+		);
+
+		doAnswer(answer) /* Behave as defined above */
+				.when(http) /* when following method is executed on {@code Http} instance */
+				.httpExecute(
+						url.capture(), /* capture url arguments passed down httpExecute to assert fallback behavior executed with valid fallback url */
+						anyString(), /* Ignore */
+						aryEq(new Param[0]), /* Ignore */
+						any(Http.RequestBody.class), /* Ignore */
+						anyBoolean(), /* Ignore */
+						any(Http.ResponseHandler.class) /* Ignore */
+				);
+
+        /* Call method with real parameters */
```

Comment:
I'll stop commenting on indentation, please review in Github and correct in one go...